File: 2020/11/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	layout := [][]string{}
	for _, line := range input {
		layout = append(layout, strings.Split(line, ""))
	}
	fmt.Println(part1(layout))
	fmt.Println(part2(layout))
}

func isValidSeat(r int, c int, layout [][]string) bool {
	return r >= 0 && c >= 0 && r <= len(layout)-1 && c <= len(layout[0])-1
}

func occupiedAdjacent(r int, c int, layout [][]string) int {
	occupied := 0
	for i := -1; i <= 1; i++ {
		for j := -1; j <= 1; j++ {
			if isValidSeat(r+i, c+j, layout) && layout[r+i][c+j] == "#" && !(i == 0 && j == 0) {
				occupied++
			}
		}
	}
	return occupied
}

func occupiedAdjacent2(r int, c int, layout [][]string) int {
	occupied := 0
	for i := -1; i <= 1; i++ {
		for j := -1; j <= 1; j++ {
			if i == 0 && j == 0 {
				continue
			}
			k := 1
			for isValidSeat(r+k*i, c+k*j, layout) {
				if layout[r+k*i][c+k*j] == "#" {
					occupied++
					break
				} else if layout[r+k*i][c+k*j] == "L" {
					break
				}
				k++
			}
		}
	}
	return occupied
}

func part1(layout [][]string) int {
	changes := 1
	for changes > 0 {
		changes = 0
		layout2 := [][]string{}
		for r, row := range layout {
			row2 := []string{}
			for c, seat := range row {
				if seat == "L" && occupiedAdjacent(r, c, layout) == 0 {
					row2 = append(row2, "#")
					changes++
				} else if seat == "#" && occupiedAdjacent(r, c, layout) >= 4 {
					row2 = append(row2, "L")
					changes++
				} else {
					row2 = append(row2, seat)
				}
			}
			layout2 = append(layout2, row2)
		}
		layout = layout2
	}
	occupied := 0
	for _, row := range layout {
		for _, seat := range row {
			if seat == "#" {
				occupied++
			}
		}
	}
	return occupied
}

func part2(layout [][]string) int {
	changes := 1
	for changes > 0 {
		changes = 0
		layout2 := [][]string{}
		for r, row := range layout {
			row2 := []string{}
			for c, seat := range row {
				if seat == "L" && occupiedAdjacent2(r, c, layout) == 0 {
					row2 = append(row2, "#")
					changes++
				} else if seat == "#" && occupiedAdjacent2(r, c, layout) >= 5 {
					row2 = append(row2, "L")
					changes++
				} else {
					row2 = append(row2, seat)
				}
			}
			layout2 = append(layout2, row2)
		}
		layout = layout2
	}
	occupied := 0
	for _, row := range layout {
		for _, seat := range row {
			if seat == "#" {
				occupied++
			}
		}
	}
	return occupied
}

File: 2020/11/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split(""));

const width = input[0].length;
const height = input.length;

class Seat {
  constructor(r, c, occupied, seatMap) {
    this.r = r;
    this.c = c;
    this.occupied = occupied;
    this.seatMap = seatMap;
    this.adjacentOccupied = 0;
  }

  evalStatus() {
    let changed = false;
    if (this.occupied && this.adjacentOccupied >= 4) {
      this.occupied = false;
      changed = true;
    } else if (!this.occupied && this.adjacentOccupied === 0) {
      this.occupied = true;
      changed = true;
    }
    this.adjacentOccupied = 0;
    return changed;
  }

  evalSurroundingSeats() {
    if (this.occupied) {
      let count = 0;
      for (let r = this.r - 1; r <= this.r + 1; r++) {
        for (let c = this.c - 1; c <= this.c + 1; c++) {
          if (r === this.r && c === this.c) continue;
          if (this.seatMap.has(`${r},${c}`)) {
            this.seatMap.get(`${r},${c}`).adjacentOccupied++;
          }
        }
      }
    }
  }

  evalStatus2() {
    let changed = false;
    if (this.occupied && this.adjacentOccupied >= 5) {
      this.occupied = false;
      changed = true;
    } else if (!this.occupied && this.adjacentOccupied === 0) {
      this.occupied = true;
      changed = true;
    }
    this.adjacentOccupied = 0;
    return changed;
  }

  evalSurroundingSeats2() {
    if (this.occupied) {
      let count = 0;
      for (let dx = -1; dx < 2; dx++) {
        for (let dy = -1; dy < 2; dy++) {
          if (dx === 0 && dy === 0) continue;
          let i = 1;
          while (true) {
            const r = this.r + dx * i;
            const c = this.c + dy * i;
            if (r < 0 || r >= height || c < 0 || c >= width) break;
            if (this.seatMap.has(`${r},${c}`)) {
              this.seatMap.get(`${r},${c}`).adjacentOccupied++;
              break;
            }
            i++;
          }
        }
      }
    }
  }
}

function parseInput(input) {
  const seatMap = new Map();
  for (let r = 0; r < height; r++) {
    for (let c = 0; c < width; c++) {
      if (input[r][c] === ".") continue;
      seatMap.set(`${r},${c}`, new Seat(r, c, input[r][c] === "#", seatMap));
    }
  }
  return seatMap;
}

function part1() {
  const seatMap = parseInput(input);
  while (true) {
    let changes = 0;

    for (const seat of seatMap.values()) {
      seat.evalSurroundingSeats();
    }
    for (const seat of seatMap.values()) {
      changes += seat.evalStatus();
    }

    if (changes === 0) {
      return Array.from(seatMap.values()).reduce((a, b) => a + b.occupied, 0);
    }
  }
}

console.log(part1());

function part2() {
  const seatMap = parseInput(input);
  while (true) {
    let changes = 0;

    for (const seat of seatMap.values()) {
      seat.evalSurroundingSeats2();
    }
    for (const seat of seatMap.values()) {
      changes += seat.evalStatus2();
    }

    if (changes === 0) {
      return Array.from(seatMap.values()).reduce((a, b) => a + b.occupied, 0);
    }
  }
}

console.log(part2());
